tensorflow/contrib/layers/kernels/layer_norm_fused_op_gpu.cu.cc

```diff
@@ -0,0 +1,1455 @@
+    mu = static_cast<T>(0.0f);
+    rstd = static_cast<T>(0.0f);
+
+    const int thread_id =
+        (bId * slice_per_block + slice_id) * in_depth + thread_slice_id;
+    // const T inp = 0;
+    const T inp = get_value<T>(input + thread_id, thread_slice_id, in_depth);
+
+    mu += inp * i_n;
+
+    for (int mask = slice_size / 2; mask > 0; mask /= 2) {
+      mu += __shfl_xor(mu, mask);
+    }
+
+    if (thread_slice_id < in_depth) rstd += (inp - mu) * (inp - mu);
+
+    for (int mask = slice_size / 2; mask > 0; mask /= 2) {
+      rstd += __shfl_xor(rstd, mask);
+    }
+
+    rstd = rsqrt(rstd * i_n + epsilon);
+
+    if (thread_slice_id < in_depth && thread_id < n_inputs)
+      output[thread_id] = (inp - mu) * rstd + _beta;
+  }
+}
+
+template <typename T, int mult>
+__global__ void LayerNormFusedGPUKernel(const LayerNormFusedArgs args,
+                                        const T* __restrict__ input,
+                                        const T* __restrict__ gamma,
+                                        const T* __restrict__ beta,
+                                        T* __restrict__ output,
+                                        const int num_blocks) {
+  const int in_depth = args.depth;
+  const int slice_size = args.slice_size;
+  const int n_inputs = args.n_inputs;
+  const T epsilon = args.epsilon;
+
+  const int thread_warp_id = threadIdx.x % warpSize;
+
+  const int thread_slice_id = threadIdx.x % slice_size;
+
+  extern __shared__ __align__(sizeof(T)) unsigned char my_smem[];
+  T* mean_cache = (T*)my_smem;
+  T* std_cache = (T*)&my_smem[sizeof(T)];
+
+  const T i_n = static_cast<T>(1.0f) / static_cast<T>(in_depth);
+  T inp[mult];
+  T _gamma[mult];
+  T _beta[mult];
+  int thread_id[mult];
+
+  T sum;
+  T sqSum;
+  T mu;
+  T rstd;
+
+  UNROLL for (int m = 0; m < mult; m++) {
+    _gamma[m] = get_value<T>(gamma + thread_slice_id + m * blockDim.x,
+                             thread_slice_id + m * blockDim.x, in_depth);
+    _beta[m] = get_value<T>(beta + thread_slice_id + m * blockDim.x,
+                            thread_slice_id + m * blockDim.x, in_depth);
+  }
+
+  for (int bId = blockIdx.x; bId < num_blocks; bId += gridDim.x) {
+    sum = static_cast<T>(0.0f);
+    sqSum = static_cast<T>(0.0f);
+
+    if (thread_slice_id == 0) {
+      mean_cache[0] = static_cast<T>(0.0f);
+      std_cache[0] = static_cast<T>(0.0f);
+    }
+    __syncthreads();
+
+    UNROLL for (int m = 0; m < mult; m++) {
+      thread_id[m] = bId * in_depth + m * blockDim.x + thread_slice_id;
+    }
+
+    UNROLL for (int m = 0; m < mult; m++) {
+      // inp[m] = __ldg(input+thread_id[m]);
+      inp[m] = get_value<T>(input + thread_id[m],
+                            thread_slice_id + m * blockDim.x, in_depth);
+      // if(blockIdx.x==0&&threadIdx.x==0)
+      // printf("m:%d,bId:%d,tId:%d,inp:%f\n",m,bId,thread_id[m],inp[m]);
+    }
+
+    UNROLL for (int m = 0; m < mult; m++) { sum += inp[m] * i_n; }
```

Comment:
I did this so CUDA can use fused multiply add (FMA) instructoins instead of adding and then multiply. It seems like this is the recommended approach from [CUDA's documentation](http://docs.nvidia.com/cuda/floating-point/#fused-multiply-add-fma), since it reduces rounding steps. Also according to the [programming guide](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#maximize-instruction-throughput) add and multiply-add actually take same amount of time.